File: tools/tf-types-ts/src/core/delegation.ts
import type { Constraint, DelegationLink } from "../generated/_common.js";
import { intersectConstraints } from "./capability.js";

export interface WalkResult {
  readonly valid: boolean;
  readonly effective: Constraint[];
  readonly expiredAt?: string;
  readonly brokenStep?: number;
  readonly reason?: string;
}

/**
 * Walk a delegation chain left-to-right (root at index 0) at time `now`.
 *
 * - If any link's `expires_at` is in the past, the chain is invalid from
 *   that step onward.
 * - Redelegation rules are enforced: a step that disallows redelegation
 *   breaks the chain at the *next* step.
 * - The effective constraint set is the intersection of all step constraints.
 */
export function walkChain(chain: readonly DelegationLink[], now: string): WalkResult {
  let effective: Constraint[] = [];
  let allowRedelegation = true;
  let maxDepthRemaining = Number.POSITIVE_INFINITY;

  for (let i = 0; i < chain.length; i++) {
    const step = chain[i]!;
    if (i > 0) {
      if (!allowRedelegation) {
        return {
          valid: false,
          effective,
          brokenStep: i,
          reason: `step ${i - 1} disallows redelegation`,
        };
      }
      if (maxDepthRemaining <= 0) {
        return {
          valid: false,
          effective,
          brokenStep: i,
          reason: `max_depth exceeded at step ${i}`,
        };
      }
      maxDepthRemaining -= 1;
    }
    if (step.expires_at && step.expires_at < now) {
      return {
        valid: false,
        effective,
        brokenStep: i,
        expiredAt: step.expires_at,
        reason: `step ${i} expired at ${step.expires_at}`,
      };
    }
    if (step.constraints && step.constraints.length > 0) {
      effective = intersectConstraints(effective, step.constraints);
    }
    const redelegation = step.redelegation as
      | { allowed?: unknown; max_depth?: unknown }
      | undefined;
    if (redelegation) {
      allowRedelegation = redelegation.allowed === true;
      if (typeof redelegation.max_depth === "number") {
        maxDepthRemaining = Math.min(maxDepthRemaining, redelegation.max_depth);
      }
    } else {
      allowRedelegation = true;
    }
  }

  return { valid: true, effective };
}

File: crates/tf-types/src/delegation.rs
//! Delegation chain walker — mirrors
//! `tools/tf-types-ts/src/core/delegation.ts`.

use crate::capability::intersect_constraints;
use crate::generated::common::{Constraint, DelegationLink};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct WalkResult {
    pub valid: bool,
    pub effective: Vec<Constraint>,
    pub expired_at: Option<String>,
    pub broken_step: Option<usize>,
    pub reason: Option<String>,
}

pub fn walk_chain(chain: &[DelegationLink], now: &str) -> WalkResult {
    let mut effective: Vec<Constraint> = Vec::new();
    let mut allow_redelegation = true;
    let mut max_depth_remaining: i64 = i64::MAX;

    for (i, step) in chain.iter().enumerate() {
        if i > 0 {
            if !allow_redelegation {
                return WalkResult {
                    valid: false,
                    effective,
                    expired_at: None,
                    broken_step: Some(i),
                    reason: Some(format!("step {} disallows redelegation", i - 1)),
                };
            }
            if max_depth_remaining <= 0 {
                return WalkResult {
                    valid: false,
                    effective,
                    expired_at: None,
                    broken_step: Some(i),
                    reason: Some(format!("max_depth exceeded at step {}", i)),
                };
            }
            max_depth_remaining -= 1;
        }

        if let Some(expires_at) = &step.expires_at {
            if expires_at.as_str() < now {
                return WalkResult {
                    valid: false,
                    effective,
                    expired_at: Some(expires_at.clone()),
                    broken_step: Some(i),
                    reason: Some(format!("step {} expired at {}", i, expires_at)),
                };
            }
        }

        if let Some(constraints) = &step.constraints {
            if !constraints.is_empty() {
                effective = intersect_constraints(&effective, constraints);
            }
        }

        if let Some(redelegation) = &step.redelegation {
            allow_redelegation = redelegation.allowed;
            if let Some(d) = redelegation.max_depth {
                max_depth_remaining = max_depth_remaining.min(d);
            }
        } else {
            allow_redelegation = true;
        }
    }

    WalkResult {
        valid: true,
        effective,
        expired_at: None,
        broken_step: None,
        reason: None,
    }
}
